hamilton/function_modifiers/recursive.py

```diff
@@ -600,3 +603,292 @@ def required_config(self) -> Optional[List[str]]:
+        )
+        def final_df(initial_df: pd.DataFrame) -> pd.DataFrame:
+            # process, or just return unprocessed
+            ...
+
+    In this instance the ``initial_df`` would get two columns added: ``a_plus_b`` and ``a_b_average``.
+
+    The operations are applied in topological order. This allows you to
+    express the operations individually, making it easy to unit-test and reuse.
+
+    Note that the operation is "append", meaning that the columns that are selected are appended
+    onto the dataframe.
+
+    If the function takes multiple dataframes, the dataframe input to process will always be
+    the first argument. This will be passed to the subdag, transformed, and passed back to the function.
+    This follows the hamilton rule of reference by parameter name. To demonstarte this, in the code
+    above, the dataframe that is passed to the subdag is `initial_df`. That is transformed
+    by the subdag, and then returned as the final dataframe.
+
+    You can read it as:
+
+    "final_df is a function that transforms the upstream dataframe initial_df, running the transformations
+    from my_module. It starts with the columns a_from_df and b_from_df, and then adds the columns
+    a, b, and a_plus_b to the dataframe. It then returns the dataframe, and does some processing on it."
+
+    In case you need more flexibility you can alternatively use ``pass_dataframe_as``, for example,
+
+    .. code-block:: python
+
+            # with_columns_module.py
+            def a_from_df(initial_df: pd.Series) -> pd.Series:
+                return initial_df["a_from_df"] / 100
+
+        def b_from_df(initial_df: pd.Series) -> pd.Series:
+                return initial_df["b_from_df"] / 100
+
+
+            # the with_columns call
+            @with_columns(
+                *[my_module],
+                *[a_from_df],
+                columns_to_pass=["a_from_df", "b_from_df"],
+                select=["a_from_df", "b_from_df", "a", "b", "a_plus_b", "a_b_average"],
+            )
+            def final_df(initial_df: pd.DataFrame) -> pd.DataFrame:
+                # process, or just return unprocessed
+                ...
+
+    the above would output a dataframe where the two columns ``a_from_df`` and ``b_from_df`` get
+    overwritten.
+    """
+
+    @staticmethod
+    def _check_for_duplicates(nodes_: List[node.Node]) -> bool:
+        "Ensures that we don't run into name clashing of columns and group operations."
+        node_counter = defaultdict(int)
+        for node_ in nodes_:
+            node_counter[node_.name] += 1
+            if node_counter[node_.name] > 1:
+                return True
+        return False
+
+    def __init__(
+        self,
+        *load_from: Union[Callable, ModuleType],
+        columns_to_pass: Union[str, List[str]] = None,
+        pass_dataframe_as: str = None,
+        select: Union[str, List[str]] = None,
```

Comment:
So I think we should keep this consistenct with the `h_spark` `with_columns` -- maybe we add a single string ability for that? Or we remove it for this?

Reply:
Will make it consistent with spark -- in case this becomes a pain point can always add it later on.